### ISSUE TITLE
Types: '@stylistic/eslint-plugin' is not assignable to 'ESLint.Plugin'

### ISSUE THREAD

**Issue by JstnMcBrd** [eslint-stylistic/eslint-stylistic#398]
### Validations

- [X] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Bug

I am trying to create a type-safe ESLint config using `@stylistic/eslint-plugin`. But it appears the types are not formatted correctly.

Here is a simplified version of what I have:
```ts
import stylistic from '@stylistic/eslint-plugin';
import type { Linter } from 'eslint';

export function config(): Linter.FlatConfig[] {
    return [
        {
            plugins: {
                // Throws type error
                // The type of 'stylistic' is not compatible with 'ESLint.Plugin'
                '@stylistic': stylistic,
            },
            // ...
        },
    ];
}
```

Here is the full error message:
```
Type 'typeof import(".../@stylistic/eslint-plugin/dist/dts/index")' is not assignable to type 'Plugin'.
  Types of property 'configs' are incompatible.
    Type '{ 'disable-legacy': FlatConfig<RulesRecord>; customize: { (options: StylisticCustomizeOptions<false>): BaseConfig<RulesRecord, RulesRecord>; (options?: StylisticCustomizeOptions<...> | undefined): FlatConfig<...>; }; ... 4 more ...; 'recommended-legacy': BaseConfig<...>; }' is not assignable to type 'Record<string, FlatConfig<RulesRecord> | FlatConfig<RulesRecord>[] | ConfigData<RulesRecord>>'.
      Property 'customize' is incompatible with index signature.
        Type '{ (options: StylisticCustomizeOptions<false>): BaseConfig<RulesRecord, RulesRecord>; (options?: StylisticCustomizeOptions<true> | undefined): FlatConfig<...>; }' is not assignable to type 'FlatConfig<RulesRecord> | FlatConfig<RulesRecord>[] | ConfigData<RulesRecord>'.
```

I have created a simple reproduction to demonstrate: https://stackblitz.com/edit/github-dbnoms

### Cause

It appears that in [eslint-plugin/configs/index.ts](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/packages/eslint-plugin/configs/index.ts#L21), you have added the `customize()` factory method to the list of configurations. But ESLint's types require that the `ESLint.Plugin["configs"]` property must be of type `Record<string, ConfigData | Linter.FlatConfig | Linter.FlatConfig[]>` - the values must be config objects, not methods.

The type of `plugin.configs` is defined in a separate .dts file - [eslint-plugin/dts/index](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/packages/eslint-plugin/dts/index.ts#L14) - which uses the exported type value of `eslint-plugin/configs/index.ts`, which sends the issue to my type-checker.

### Proposed solutions

It should be possible to define the config's type in such a way that keeps `customize()` but also tells TypeScript that configs is compatible with `ESLint.Plugin["configs"]`.

Locally, I was able to edit the dts files in `node_modules` and defined `config` as an extension of `ESLint.Plugin["configs"]` and the type error went away.

So in [eslint-plugin/dts/index](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/packages/eslint-plugin/dts/index.ts#L14), you could do this:
```diff
- configs: typeof configs
+ configs: ESLint.Plugin["configs"] & typeof configs
```

Of course, you may know of a better solution - I am not familiar with how the TypeScript build process works for this repository. For example, I don't know why there are custom .dts files instead of letting TypeScript generate them. Whatever the reason, perhaps it will influence what solution you choose.

### Reproduction

https://stackblitz.com/edit/github-dbnoms

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests

**Comment by silverwind** [comment]
Noticed the same type issue in this and also other plugins like `@typescript-eslint`. Working around it with a custom module definition currently:

```ts
declare module "@stylistic/eslint-plugin-js" {
  import type {Eslint} from "eslint";
  const plugin: Eslint.Plugin;
  export = plugin;
}

**Comment by JstnMcBrd** [comment]
This issue appears to not actually be fixed. I will return to it later.

**Comment by JstnMcBrd** [comment]
Can confirm, the issue is now fixed with `2.2.1`.

**Comment by silverwind** [comment]
Seems the fix is missing at least for these packages:

- `@stylistic/eslint-plugin-js`
- `@stylistic/eslint-plugin-jsx`
- `@stylistic/eslint-plugin-ts`

**Comment by JstnMcBrd** [comment]
> Seems the fix is still missing for the "sub-packages":

Hmm, are you encountering the same error with those packages? I haven't tested them, but none of them contain the `customize` method in their configs (`eslint-plugin-plus` doesn't even contain a `configs` object), so casting the configs type shouldn't be necessary.

**Comment by silverwind** [comment]
Yeah, I'm using the first two (js,jsx) and both show the error after upgrading to 2.2.1. The type definition is:

```ts
declare const plugin: {
  rules: Rules
  configs: {
    /**
     * Disable all legacy rules from `eslint`
     *
     * This config works for both flat and legacy config format
     */
    'disable-legacy': Linter.FlatConfig
    /**
     * Enable all rules, in Flat Config Format
     */
    'all-flat': Linter.FlatConfig
    /**
     * Enable all rules, in Legacy Config Format
     */
    'all-extends': Linter.BaseConfig
  }
}

export default plugin
```

**Comment by JstnMcBrd** [comment]
Yeah, that type definition does not contain the `configs.customize()` method that prevents the plugin from being compatible with `ESLint.Plugin`, which is what this bug report was about.

I tried importing `eslint-plugin-js` myself, and I got this error:

```
Type 'typeof import(".../@stylistic/eslint-plugin-js/dts/index")' has no properties in common with type 'Plugin'.
```

Which is similar, but **not** the same error as this bug report. I played around with the types for a bit and couldn't figure out the root of the issue.

It looks like there's a different problem going on here. I recommend opening a new issue and providing a full, detailed report with a minimal reproduction for the maintainers.

**Comment by silverwind** [comment]
Yeah, likely a different issue. Yes that are the errors I'm seeing:

````
index.ts:74:5 - error TS2559: Type 'typeof import("node_modules/@stylistic/eslint-plugin-js/dts/index")' has no properties in common with type 'Plugin'.

74     "@stylistic/js": stylisticJs,
       ~~~~~~~~~~~~~~~

index.ts:75:5 - error TS2559: Type 'typeof import("node_modules/@stylistic/eslint-plugin-jsx/dts/index")' has no properties in common with type 'Plugin'.

75     "@stylistic/jsx": stylisticJsx,
       ~~~~~~~~~~~~~~~~
```
